### PR TITLE
Add allowedInstallationIds list to each github app

### PR DESCRIPTION
**Pull request by punkle**
## Hey, I just made a Pull Request!

This allows a single instance of backstage to optionally limit the set of
github app installations that may be used by backstage.

Previously, if you had github app installations for tenant1 and tenant2
there was nothing stopping the first from accessing resources of the
second.

The default behaviour of the GithubCredentialsProvider remains the same.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by punkle** [comment]
This feature might be useful for a team who are deploying a test and a production instance of backstage. They might have a github app installed in an organization for testing and an organization for production.

In this case they would include the `allowedInstallationIds` to limit the github app installations for each backstage deployment.

**Comment by Rugvip** [comment]
Could another option be to limit by the org name instead?

**Comment by punkle** [comment]
@jhaals I can see the confusion alright, and I could of course be straying outside of a standard backstage use case here. We are hosting a SaaS version of backstage. Bear with me while I explain.

A single "github app" can have multiple installations on different organizations. The contents of each github organization's data is private to the member of that organization. This feature allows a single backstage instance to be limited to a sub set of the installations so that data of each organization's data are not leaked to other users.

Here are some cases for this:
- For testing purposes, you make two github app installations, one installation is on your production github organization and the second is on a github organization (or your personal account) you create for testing. With this feature you can run more than one backstage, and have each of them use different github app installations.
- For production use, you might have github organizations for different depts within an development organization. You might choose to have different backstages for each organization.
- For the SaaS use case, we have one github app for all of our customers. We must ensure each customer backstage instance has no ability to use another installations credentials. This feature allows us to configure this.

@Rugvip yes, I believe you could use the org name - actually I think you would need to use the org "login" field, but that's pedantic. I am not sure if the org name is unique.

**Comment by Rugvip** [comment]
@punkle So the use-case is clear and it's pretty much just down to implementation. It's a bit unfortunate that the GitHub credentials provider is itself not provided so you could create a custom one, but that's where we are.

I'd say a refactor to have it done by org name should be enough, since installation ID is awkward because of the chicken and egg problem and I don't think you want that either. And yeah they call the org name login in the API. Definitely don't filter by the display name :grin:

**Comment by punkle** [comment]
@Rugvip I changed it to filter on the installation owner rather than the id. Hope that works for you.

File: docs/plugins/github-apps.md [hunk 1]
```diff
@@ -89,6 +89,26 @@ integrations:
...
 ```
 
+### Limiting the GitHub App installations
+
+If you want to limit the GitHub app installations visible to backstage you may
+optionally include the `allowedInstallationOwners` option.
+
+```yaml
+appId: 1
+allowedInstallationOwners: [1234]
```
Comment (Rugvip) [inline docs/plugins/github-apps.md:99]: Update this to be some dummy name instead? 😅

File: packages/integration/src/github/config.ts [hunk 2]
```diff
@@ -93,6 +93,10 @@ export type GithubAppConfig = {
    * Client secrets can be generated at https://github.com/organizations/$org/settings/apps/$AppName
    */
   clientSecret: string;
+  /**
+   * List of installation owners allowed to be used by this backstage https://github.com/app/installations/$InstallationId
```
Comment (Rugvip) [inline packages/integration/src/github/config.ts:97]: This leads to a 404

**Review by Rugvip — COMMENTED** [review]
Yep, looks alright, just some doc updates 👍

File: packages/integration/src/github/config.ts (hunk 2 above)
Comment (punkle) [reply packages/integration/src/github/config.ts]: The link was wrong. I also added a bit of detail in this code comment. Specifically, the App installation list can only retrieved by the Application via its login token signed by it's private key. I pointed at the relevant docs.

Let me know how it reads. Happy to review again if needed. :)

File: docs/plugins/github-apps.md (hunk 1 above)
Comment (punkle) [reply docs/plugins/github-apps.md]: Done

Comment (punkle) [reply docs/plugins/github-apps.md]: Done

**Comment by punkle** [comment]
The test failures seem unrelated. Perhaps I need to update the pull request source branch?

**Comment by Rugvip** [comment]
@punkle I believe it was fixed by https://github.com/backstage/backstage/pull/6740, so try bringing in the latest master

**Comment by punkle** [comment]
Done thanks @Rugvip

**Review by Rugvip — APPROVED** [review]
🎉 , :shipit:

**Pull request merged**